R Code pulled from https://www.statmethods.net/advgraphs/ggplot2.html

In [1]:
install.packages("ISLR")

Installing package into ‘/srv/rlibs’
(as ‘lib’ is unspecified)


In [6]:
library(ISLR)
set.seed(1) 

In [19]:
names(Auto)
dim(Auto)
Auto

[1] "mpg"          "cylinders"    "displacement" "horsepower"   "weight"      
[6] "acceleration" "year"         "origin"       "name"

[1] 392   9

mpg cylinders displacement horsepower weight acceleration year origin
1   18  8         307          130        3504   12.0         70   1     
2   15  8         350          165        3693   11.5         70   1     
3   18  8         318          150        3436   11.0         70   1     
4   16  8         304          150        3433   12.0         70   1     
5   17  8         302          140        3449   10.5         70   1     
6   15  8         429          198        4341   10.0         70   1     
7   14  8         454          220        4354    9.0         70   1     
8   14  8         440          215        4312    8.5         70   1     
9   14  8         455          225        4425   10.0         70   1     
10  15  8         390          190        3850    8.5         70   1     
11  15  8         383          170        3563   10.0         70   1     
12  14  8         340          160        3609    8.0         70   1     
13  15  8         400          150        3761    9.5         70   1     
14  14  8         455          225        3086   10.0         70   1     
15  24  4         113           95        2372   15.0         70   3     
16  22  6         198           95        2833   15.5         70   1     
17  18  6         199           97        2774   15.5         70   1     
18  21  6         200           85        2587   16.0         70   1     
19  27  4          97           88        2130   14.5         70   3     
20  26  4          97           46        1835   20.5         70   2     
21  25  4         110           87        2672   17.5         70   2     
22  24  4         107           90        2430   14.5         70   2     
23  25  4         104           95        2375   17.5         70   2     
24  26  4         121          113        2234   12.5         70   2     
25  21  6         199           90        2648   15.0         70   1     
26  10  8         360          215        4615   14.0         70   1     
27  10  8         307          200        4376   15.0         70   1     
28  11  8         318          210        4382   13.5         70   1     
29   9  8         304          193        4732   18.5         70   1     
30  27  4          97           88        2130   14.5         71   3     
⋮   ⋮   ⋮         ⋮            ⋮          ⋮      ⋮            ⋮    ⋮     
368 28  4         112           88        2605   19.6         82   1     
369 27  4         112           88        2640   18.6         82   1     
370 34  4         112           88        2395   18.0         82   1     
371 31  4         112           85        2575   16.2         82   1     
372 29  4         135           84        2525   16.0         82   1     
373 27  4         151           90        2735   18.0         82   1     
374 24  4         140           92        2865   16.4         82   1     
375 36  4         105           74        1980   15.3         82   2     
376 37  4          91           68        2025   18.2         82   3     
377 31  4          91           68        1970   17.6         82   3     
378 38  4         105           63        2125   14.7         82   1     
379 36  4          98           70        2125   17.3         82   1     
380 36  4         120           88        2160   14.5         82   3     
381 36  4         107           75        2205   14.5         82   3     
382 34  4         108           70        2245   16.9         82   3     
383 38  4          91           67        1965   15.0         82   3     
384 32  4          91           67        1965   15.7         82   3     
385 38  4          91           67        1995   16.2         82   3     
386 25  6         181          110        2945   16.4         82   1     
387 38  6         262           85        3015   17.0         82   1     
388 26  4         156           92        2585   14.5         82   1     
389 22  6         232          112        2835   14.7         82   1     
390 32  

In [21]:
train=sample(392,196)
train

[1] 363 199  61 136 257 121 389  57 254  71 365 343 359 275 140 295   5 353
 [19] 372 134 279 377 262 176 183 114 255 301 159 187 241  52 125 146  31 333
 [37] 299 313 332  26 373 189  37 280 258  19 167 318  15 101 172 208  90 144
 [55] 124 345  42  24 323 148 123 385  18 217 190 334 197  22  53 154   1 142
 [73]  84 300 228 361 151 218 145 356 223 338 350  75 266 352 153 212 232  48
 [91] 388 285 177 316 331 322 260  40   7 276  86 303 116 133 126 149 243  16
[109] 158 195 186 392 289 271 112 236 209 147 240 128   3 198 194  51 174 314
[127] 340 170 219 281  92  34 381 362 341 283 246 384 155   8  85  70  30  11
[145] 370 320  43 306  98 233 364  80 291  28 238 308 349 173 204 307 272 210
[163] 178  88 325  12 282 263 379 250  76 162 200 245 324 378 390 206 261 237
[181]  54  20 268 110 139 252 231 152  72 193 131 253 120  83 382 105

In [16]:
lm.fit=lm(mpg~horsepower,data = Auto, subset = train)
attach(Auto)
summary(lm.fit)

The following objects are masked from Auto (pos = 3):

    acceleration, cylinders, displacement, horsepower, mpg, name,
    origin, weight, year




Call:
lm(formula = mpg ~ horsepower, data = Auto, subset = train)

Residuals:
    Min      1Q  Median      3Q     Max 
-13.698  -3.085  -0.216   2.680  16.770 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 40.340377   1.002269   40.25   <2e-16 ***
horsepower  -0.161701   0.008809  -18.36   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 4.692 on 194 degrees of freedom
Multiple R-squared:  0.6346,	Adjusted R-squared:  0.6327 
F-statistic: 336.9 on 1 and 194 DF,  p-value: < 2.2e-16


In [36]:
pred = predict(lm.fit,Auto)
pred

1         2         3         4         5         6         7         8 
19.319212 13.659667 16.085186 16.085186 17.702199  8.323525  4.766097  5.574604 
        9        10        11        12        13        14        15        16 
 3.957591  9.617136 12.851161 14.468174 16.085186  3.957591 24.978756 24.978756 
       17        18        19        20        21        22        23        24 
24.655354 26.595769 26.110665 32.902119 26.272367 25.787263 24.978756 22.068133 
       25        26        27        28        29        30        31        32 
25.787263  5.574604  8.000123  6.383110  9.132032 26.110665 25.787263 24.978756 
       34        35        36        37        38        39        40        41 
24.170250 23.361744 24.170250 26.110665 24.170250 13.659667 12.042655 15.600083 
       42        43        44        45        46        47        48        49 
16.085186 11.234148 12.851161 12.042655 22.553237 28.697886 24.170250 26.110665 
       50        51        52        53        54        55        56        57 
26.434068 25.787263 29.021288 28.051081 29.829795 29.182989 30.638301 29.021288 
       58        59        60        61        62        63        64        65 
24.978756 27.404275 31.608509 25.787263 26.434068 13.659667 12.042655 16.085186 
       66        67        68        69        70        71        72        73 
15.600083 16.085186  6.706513 15.276680 14.468174  9.617136 24.655354 16.085186 
       74        75        76        77        78        79        80        81 
19.319212 17.702199 16.085186 22.229835 28.051081 26.272367 29.182989 26.434068 
       82        83        84        85        86        87        88        89 
25.463860 24.655354 27.404275 26.110665 12.042655 16.085186 16.893693 18.187303 
       90        91        92        93        94        95        96        97 
16.085186  8.323525 16.085186 14.791576 16.085186  5.574604  3.957591 12.042655 
       98        99       100       101       102       103       104       105 
23.361744 24.170250 24.170250 26.110665 24.978756 32.902119 16.085186 13.336265 
      106       107       108       109       110       111       112       113 
12.851161 11.234148 24.170250 26.110665 28.697886 25.140458 25.787263 26.595769 
      114       115       116       117       118       119       120       121 
23.038341 25.787263 16.893693  3.149085 32.417015 28.212782 25.625561 22.229835 
      122       123       124       125       126       128       129       130 
16.085186 22.553237 20.612822 11.234148 24.978756 24.170250 24.170250 29.506392 
      131       132       133       134       135       136       137       138 
27.404275 29.829795 28.212782 24.170250 22.553237 23.361744 17.702199 16.085186 
      139       140       141       142       143       144       145       146 
16.085186 17.702199 16.085186 26.919172 29.506392 27.727678 31.931911 30.476600 
      147       148       149       150       151       152       153       154 
28.212782 28.212782 28.212782 24.655354 25.302159 29.506392 24.978756 23.361744 
      155       156       157       158       159       160       161       162 
28.697886 28.697886 12.851161 16.893693 16.085186 16.408589 22.553237 23.361744 
      163       164       165       166       167       168       169       170 
22.553237 24.978756 22.553237 22.553237 19.480913 28.212782 26.919172 24.170250 
      171       172       173       174       175       176       177       178 
27.727678 24.817055 28.859587 24.655354 24.655354 29.021288 25.787263 24.978756 
      179       180       181       182       183       184       185       186 
26.110665 24.493653 21.744731 31.770210 26.434068 27.242574 25.463860 27.565977 
      187       188       189       190       191       192       193       194 
26.919172 17.702199 16.085186 20.936225 15.761784 24.170250 23.361744 27.242574 
      195       196       197       198       199       200       201       202 
25.787263 31.931911 30.638301 29.021288

In [46]:
mean((mpg-pred)[-train]^2)

[1] 22.35896

In [40]:
lm.fit2 = lm(mpg~poly(horsepower,2), subset = train)
summary(lm.fit2)


Call:
lm(formula = mpg ~ poly(horsepower, 2), subset = train)

Residuals:
     Min       1Q   Median       3Q      Max 
-15.0771  -2.3219  -0.1393   2.1552  15.7553 

Coefficients:
                      Estimate Std. Error t value Pr(>|t|)    
(Intercept)            23.7207     0.3239  73.225  < 2e-16 ***
poly(horsepower, 2)1 -121.0354     6.6601 -18.173  < 2e-16 ***
poly(horsepower, 2)2   46.0377     6.4545   7.133 1.93e-11 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 4.529 on 193 degrees of freedom
Multiple R-squared:  0.6685,	Adjusted R-squared:  0.665 
F-statistic: 194.6 on 2 and 193 DF,  p-value: < 2.2e-16


In [42]:
pred2 = predict(lm.fit2,Auto)
pred2

1        2        3        4        5        6        7        8 
17.23343 13.66534 14.80905 14.80905 15.89275 13.18447 14.41863 14.02892 
       9       10       11       12       13       14       15       16 
14.87258 13.04406 13.41259 13.98233 14.80905 14.87258 23.94955 23.94955 
      17       18       19       20       21       22       23       24 
23.48097 26.44665 25.67054 38.64080 25.92667 25.16598 23.94955 20.10236 
      25       26       27       28       29       30       31       32 
25.16598 14.02892 13.24527 13.70346 13.07744 25.67054 25.16598 23.94955 
      34       35       36       37       38       39       40       41 
22.79737 21.70943 22.79737 25.67054 22.79737 13.66534 13.22409 14.53405 
      42       43       44       45       46       47       48       49 
14.80905 13.09984 13.41259 13.22409 20.68574 30.07707 22.79737 25.67054 
      50       51       52       53       54       55       56       57 
26.18537 25.16598 30.67414 28.91376 32.21179 30.97653 33.81369 30.67414 
      58       59       60       61       62       63       64       65 
23.94955 27.79157 35.82078 25.16598 26.18537 13.66534 13.22409 14.80905 
      66       67       68       69       70       71       72       73 
14.53405 14.80905 13.59126 14.36357 13.98233 13.04406 23.48097 14.80905 
      74       75       76       77       78       79       80       81 
17.23343 15.89275 14.80905 20.29425 28.91376 25.92667 30.97653 26.18537 
      82       83       84       85       86       87       88       89 
24.67170 23.48097 27.79157 25.67054 13.22409 14.80905 15.31878 16.26797 
      90       91       92       93       94       95       96       97 
14.80905 13.18447 14.80905 14.12711 14.80905 14.02892 14.87258 13.22409 
      98       99      100      101      102      103      104      105 
21.70943 22.79737 22.79737 25.67054 23.94955 38.64080 14.80905 13.55653 
     106      107      108      109      110      111      112      113 
13.41259 13.09984 22.79737 25.67054 30.07707 24.18770 25.16598 26.44665 
     114      115      116      117      118      119      120      121 
21.29225 25.16598 15.31878 15.39078 37.56402 29.20073 24.91755 20.29425 
     122      123      124      125      126      128      129      130 
14.80905 20.68574 18.49100 13.09984 23.94955 22.79737 22.79737 31.58902 
     131      132      133      134      135      136      137      138 
27.79157 32.21179 29.20073 22.79737 20.68574 21.70943 15.89275 14.80905 
     139      140      141      142      143      144      145      146 
14.80905 15.89275 14.80905 26.97691 31.58902 28.34752 36.51036 33.48817 
     147      148      149      150      151      152      153      154 
29.20073 29.20073 29.20073 23.48097 24.42841 31.58902 23.94955 21.70943 
     155      156      157      158      159      160      161      162 
30.07707 30.07707 13.41259 15.31878 14.80905 15.00523 20.68574 21.70943 
     163      164      165      166      167      168      169      170 
20.68574 23.94955 20.68574 20.68574 17.38163 29.20073 26.97691 22.79737 
     171      172      173      174      175      176      177      178 
28.34752 23.71397 30.37432 23.48097 23.48097 30.67414 25.16598 23.94955 
     179      180      181      182      183      184      185      186 
25.67054 23.25053 19.72629 36.16428 26.18537 27.51744 24.67170 28.06826 
     187      188      189      190      191      192      193      194 
26.97691 15.89275 14.80905 18.83109 14.62315 22.79737 21.70943 27.51744 
     195      196      197      198      199      200      201      202 
25.16598 36.51036 33.81369 30.67414 36.16428 22.79737 28.34752 20.68574 
     203      204      205      206      207      208      209      210 
23.94955 30.37432 30.67414 29.20073 30.07707 22.35448 14.80905 25.67054 
     211      212      213      214      215      216      217      218 
21.08751 18.83109 13.09984 15.31878 17.23343 14.80905 31.28149 27.79157 
     219      220      221      222      223      224      225   

In [45]:
mean(((mpg-predict(lm.fit2,Auto))[-train])^2)

[1] 17.95

In [48]:
lm.fit3 = lm(mpg~poly(horsepower,3), subset = train)
pred3 = predict(lm.fit3, Auto)
pred3

1        2        3        4        5        6        7        8 
17.43111 14.36596 15.36862 15.36862 16.28469 13.30627 13.08542 13.12107 
       9       10       11       12       13       14       15       16 
13.05212 13.45528 14.11699 14.65526 15.36862 13.05212 23.65628 23.65628 
      17       18       19       20       21       22       23       24 
23.19477 26.18104 25.38473 39.93518 25.64639 24.87260 23.65628 19.98683 
      25       26       27       28       29       30       31       32 
24.87260 13.12107 13.27659 13.16251 13.39325 25.38473 24.87260 23.65628 
      34       35       36       37       38       39       40       41 
22.52862 21.48616 22.52862 25.38473 22.52862 14.36596 13.90489 15.13458 
      42       43       44       45       46       47       48       49 
15.36862 13.72621 14.11699 13.90489 20.52545 30.03883 22.52862 25.38473 
      50       51       52       53       54       55       56       57 
25.91182 24.87260 30.69354 28.77936 32.40495 31.02724 34.22573 30.69354 
      58       59       60       61       62       63       64       65 
23.65628 27.58504 36.55990 24.87260 25.91182 14.36596 13.90489 15.36862 
      66       67       68       69       70       71       72       73 
15.13458 15.36862 13.18148 14.98832 14.65526 13.45528 23.19477 15.36862 
      74       75       76       77       78       79       80       81 
17.43111 16.28469 15.36862 20.16329 28.77936 25.64639 31.02724 25.91182 
      82       83       84       85       86       87       88       89 
24.37521 23.19477 27.58504 25.38473 13.90489 15.36862 15.79959 16.60318 
      90       91       92       93       94       95       96       97 
15.36862 13.30627 15.36862 14.78304 15.36862 13.12107 13.05212 13.90489 
      98       99      100      101      102      103      104      105 
21.48616 22.52862 22.52862 25.38473 23.65628 39.93518 15.36862 14.26173 
     106      107      108      109      110      111      112      113 
14.11699 13.72621 22.52862 25.38473 30.03883 23.89234 24.87260 26.18104 
     114      115      116      117      118      119      120      121 
21.09226 24.87260 15.79959 13.01770 38.63346 29.08805 24.62208 20.16329 
     122      123      124      125      126      128      129      130 
15.36862 20.52545 18.53265 13.72621 23.65628 22.52862 22.52862 31.70745 
     131      132      133      134      135      136      137      138 
27.58504 32.40495 29.08805 22.52862 20.52545 21.48616 16.28469 15.36862 
     139      140      141      142      143      144      145      146 
15.36862 16.28469 15.36862 26.73098 31.70745 28.17417 37.37509 33.85265 
     147      148      149      150      151      152      153      154 
29.08805 29.08805 29.08805 23.19477 24.13198 31.70745 23.65628 21.48616 
     155      156      157      158      159      160      161      162 
30.03883 30.03883 14.11699 15.79959 15.36862 15.53473 20.52545 21.48616 
     163      164      165      166      167      168      169      170 
20.52545 23.65628 20.52545 20.52545 17.55949 29.08805 26.73098 22.52862 
     171      172      173      174      175      176      177      178 
28.17417 23.42377 30.36408 23.19477 23.19477 30.69354 24.87260 23.65628 
     179      180      181      182      183      184      185      186 
25.38473 22.96926 19.64305 36.96514 25.91182 27.29643 24.37521 27.87761 
     187      188      189      190      191      192      193      194 
26.73098 16.28469 15.36862 18.83549 15.21061 22.52862 21.48616 27.29643 
     195      196      197      198      199      200      201      202 
24.87260 37.37509 34.22573 30.69354 36.96514 22.52862 28.17417 20.52545 
     203      204      205      206      207      208      209      210 
23.65628 30.36408 30.69354 29.08805 30.03883 22.10160 15.36862 25.38473 
     211      212      213      214      215      216      217      218 
20.90014 18.83549 13.72621 15.79959 17.43111 15.36862 31.36520 27.58504 
     219      220      221      222      223      224      225   

In [50]:
mean((mpg-pred3)[-train]^2)

[1] 18.16543

As we see the error of model when cubic term is included is more than the error when only upto square term was included.
So now we will see the p-value of cubic term to decide whether it is a significant variable in deciding the mpg.  

In [51]:
summary(lm.fit3)


Call:
lm(formula = mpg ~ poly(horsepower, 3), subset = train)

Residuals:
     Min       1Q   Median       3Q      Max 
-15.0388  -2.3747  -0.3686   2.2702  15.5179 

Coefficients:
                      Estimate Std. Error t value Pr(>|t|)    
(Intercept)            23.7367     0.3235  73.374  < 2e-16 ***
poly(horsepower, 3)1 -120.6662     6.6522 -18.139  < 2e-16 ***
poly(horsepower, 3)2   45.3455     6.4620   7.017 3.78e-11 ***
poly(horsepower, 3)3   -8.4684     6.3277  -1.338    0.182    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 4.519 on 192 degrees of freedom
Multiple R-squared:  0.6715,	Adjusted R-squared:  0.6664 
F-statistic: 130.8 on 3 and 192 DF,  p-value: < 2.2e-16


As we see the p-value of cubic term is 0.182 which is quite large so it is excluded from the model

In [53]:
glm.fit = glm(mpg~horsepower, data = Auto)
summary(glm.fit)


Call:
glm(formula = mpg ~ horsepower, data = Auto)

Deviance Residuals: 
     Min        1Q    Median        3Q       Max  
-13.5710   -3.2592   -0.3435    2.7630   16.9240  

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 39.935861   0.717499   55.66   <2e-16 ***
horsepower  -0.157845   0.006446  -24.49   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for gaussian family taken to be 24.06645)

    Null deviance: 23819.0  on 391  degrees of freedom
Residual deviance:  9385.9  on 390  degrees of freedom
AIC: 2363.3

Number of Fisher Scoring iterations: 2


In [55]:
library(boot)
cv.error=cv.glm(Auto,glm.fit)
cv.error$delta

[1] 24.23151 24.23114

In [59]:
cv.error=rep(0,5)
for(i in 1:5){
    glm.fit = glm(mpg~poly(horsepower,i))
    cv.error[i] = cv.glm(Auto,glm.fit)$delta[1]
    }
cv.error

[1] 24.23151 19.24821 19.33498 19.42443 19.03321

In [61]:
set.seed(17)
cv.error.10 = rep(0,10)
for(i in 1:10){
     glm.fit = glm(mpg~poly(horsepower,i))
    cv.error.10[i] = cv.glm(Auto,glm.fit,K=10)$delta[1]
}
cv.error.10

[1] 24.20520 19.18924 19.30662 19.33799 18.87911 19.02103 18.89609 19.71201
 [9] 18.95140 19.50196